runtime/tsan.c

```diff
@@ -98,7 +98,99 @@
+   to C11 accesses.
+
+   For each type of OCaml memory access on the left, we signal to TSan the
+   operations on the right:
+
+   OCaml access      | C11 equivalent          | TSan view
+   ------------------|-------------------------|--------------------------------
+   Atomic load       | fence(acquire)          | __tsan_atomic64_load(seq_cst)
+                     | atomic_load(seq_cst)    |
+   ------------------|-------------------------|--------------------------------
+   Atomic store      | fence(acquire)          | __tsan_atomic_exchange(seq_cst)
+                     | atomic_exchange(seq_cst)|
+                     | fence(release)          |
+   ------------------|-------------------------|--------------------------------
+   Non-atomic load   | atomic_load(relaxed)    | __tsan_read8()
+   ------------------|-------------------------|--------------------------------
+   Non-atomic store  | fence(acquire)          | __tsan_write8()
+   (assignment, int) | atomic_store(release)   |
+   ------------------|-------------------------|--------------------------------
+   Non-atomic store  | fence(acquire)          | __tsan_write8()
+   (assignment, ptr) | atomic_store(release)   |
+                     | fence(release)          |
+   ------------------|-------------------------|--------------------------------
+   Non-atomic store  | plain store             | Not instrumented
+   (initialization)  |                         |
+   ------------------|-------------------------|--------------------------------
+   Non-atomic store  | plain store             | __tsan_writeN()
+   (unaligned size)  |                         |
+   ------------------|-------------------------|--------------------------------
+
+   This mapping dictates which instrumentation calls are generated by the
+   compiler, but also how some functions of the runtime are instrumented: for
+   example, caml_modify is instrumented with a call to __tsan_write8, whereas
+   caml_initialize is purposefully not instrumented.
+
+   3.1. False negatives
+
+   There should be no false negatives, i.e., all data races (in the OCaml
+   sense) on visited code paths should be detected (modulo TSan limitations
+   such as the finite history of memory accesses).
+
+   3.2. False positives
+
+   The mapping that we use should not incur any false positives in pure OCaml
+   code, i.e., all data races reported between two accesses from OCaml are
+   true data races (in the OCaml sense of the term).
+
+   In mixed C-OCaml code, rare false positives may occur in the following cases:
+
+   - A value is initialized from C without using `caml_initialize` (allowed by
+     the FFI rules on the condition that the GC does not run between the
+     allocation and the end of initialization) and a conflicting access is made
+     from OCaml after publication to other threads. There is no data race due
+     to data dependencies, but TSan does not take data dependencies into
```

Comment:
The use of data dependency is still an open problem for FFI usage, so maybe one could write less affirmatively, “There should be no data race due to data dependency (see [MMMOC] in memory.c)”.

Reply:
[dc05f9c](https://github.com/ocaml/ocaml/pull/12694/commits/dc05f9c8d3d91cf3aec603fc90c3db20e241b2b3)